core/metacling/src/TCling.cxx

```diff
@@ -5891,11 +5869,98 @@ Int_t TCling::AutoParse(const char *cls)
+      ModulePaths.erase(last, ModulePaths.end());
+
+      cling::DynamicLibraryManager* dyLibManager = fInterpreter->getDynamicLibraryManager();
+
+      int cur_lib = 0;
+      for (auto Path : ModulePaths) {
+         StringRef DirPath(Path);
+         if (DirPath.empty() || !is_directory(DirPath) || Path == ".")
+            continue;
+
+         // Check if this is a build directory (which should contain pcm)
+         std::error_code EC;
+         bool is_build = false;
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+            std::string FileName(DirIt->path());
+            if (extension(FileName) == ".pcm") {
+               is_build = true;
+               break;
+            }
+         }
+
+         if (!is_build)
+            continue;
+
+         // Iterate over the directory
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+            std::string LibName(DirIt->path());
+            if (llvm::sys::fs::is_directory(LibName) || extension(LibName) != ".so")
+               continue;
+
+            // TCling::IsLoaded is incredibly slow!
+            if (dyLibManager->isLibraryLoaded(LibName.c_str()))
+               continue;
+
+            int_and_libtable.push_back(LibName);
+            auto SoFile = ObjectFile::createObjectFile(LibName);
+            if (SoFile) {
+               auto RealSoFile = SoFile.get().getBinary();
+               auto Symbols = RealSoFile->symbols();
+               for (auto S : Symbols) {
+                  // DO NOT load to table if symbol was weak or undefined
+                  if (S.getFlags() == SymbolRef::SF_Weak || S.getFlags() == SymbolRef::SF_Undefined) continue;
+
+                  std::string SymName = std::string(S.getName().get());
+                  // Put into map
+                  libtable.insert(std::make_pair(hash_fn(SymName), cur_lib));
+               }
+            }
+            cur_lib++;
+         }
+      }
+      fFirstRun = false;
+   }
+
+   auto key = libtable.find(hash_fn(mangled_name));
+   if (key != libtable.end()) {
+      std::string LoadedLibName = int_and_libtable[key->second];
+      if (!LoadedLibName.empty() && (gSystem->Load(LoadedLibName.c_str(), "", false) < 0))
+         std::cout << "Failed to load library" << std::endl;
+   }
+
+   void* addr = llvm::sys::DynamicLibrary::SearchForAddressOfSymbol(mangled_name.c_str());
+   //fprintf(stderr, "addr: %016lx\n", reinterpret_cast<unsigned long>(addr));
+   return addr;
+}
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Autoload a library based on a missing symbol.
 
 void* TCling::LazyFunctionCreatorAutoload(const std::string& mangled_name) {
+#ifdef R__USE_CXXMODULES
+   return LazyFunctionCreatorAutoloadForModule(mangled_name);
```

Comment:
Do we know if we can turn on this for regular ROOT and replace the current functionality? I'd like to know if our new implementation has comparable performance to the legacy one.

Reply:
In normal root release, without calling this function:
```
Processing tutorials/hsimple.C...                   
hsimple   : Real Time =   0.04 seconds Cpu Time =   0.04 seconds                                         
(TFile *) 0x55a0679b31b0                            
Processing /home/yuka/CERN/ROOT/memory.C...         
 cpu  time = 0.189830 seconds                       
 sys  time = 0.012380 seconds                       
 res  memory = 143.039 Mbytes                       
 vir  memory = 215.895 Mbytes         
```
In normal root release, with calling this function:
```
Processing tutorials/hsimple.C...
hsimple   : Real Time =   0.04 seconds Cpu Time =   0.04 seconds
(TFile *) 0x559c6c4a33f0
Processing /home/yuka/CERN/ROOT/memory.C...
 cpu  time = 0.185751 seconds
 sys  time = 0.024228 seconds
 res  memory = 142.578 Mbytes
 vir  memory = 215.895 Mbytes
```

So it's almost the same I would say

Reply:
I think we can even replace exisiting LazyFunctionCreatorAutoload because LazyFunctionCreatorAutoload is not relying onto modules at all